### DISCUSSION TITLE
Possibility to delete many objects

### DISCUSSION THREAD

**Discussion by larspalo** [GrandOrgue/GoOdf#18] [General]
After having worked with GOODF for a while now I've noticed that some tasks that can be performed rapidly in a text editor is quite cumbersome to do in GOODF. Especially if one wants to delete a substantial number of objects (in the tree).

Currently it's not possible to select multiple elements in the organ tree since the single selection is the way to navigate between different panels to create/edit elements. This could of course be changed with some considerable effort, though. Also, the popup message asking for confirmation about deleting an element and the re-setting to the parent as next active panel is troublesome, I think.

Now the question arises how to best implement multiple deletes... Should it be possible to select multiple elements in the tree (of the same parent) and for instance delete them all in some sensible way or should there be some other way (like a separate choice in the menu, or something similar, to delete multiple elements that the user can choose)? Opinions on the procedure the user has to execute?

**Comment by larspalo** [comment]
I'm leaning towards creating a pop-up menu for when a tree item is right clicked (with mouse). For the copying I omitted the menu, but now I think that it would be sensible to have it with different options depending on what kind of element is selected.

For the actual deleting UI, I'm leaning towards a similar solution as the Gui Element properties copying dialog where the user in a second stage can select which elements to remove. 

The whole procedure could then be that when an item in the organ tree is right clicked a pop up menu present a few options depending on what type of element is right clicked. One could, for removable elements, be "Delete many..." which would open a dialog, similar to the properties copying dialog, that presents all elements with the same parent and gives the user the possibility to select which ones of them to delete.

Any thoughts about that?

**Comment by larspalo** [comment]
On a second thought I think that I'll first implement a much simpler solution. Instead of selecting the parent of a removed item a sibling could be selected. Also the message box when removing elements could be shown only the first time the user removes that type of item... More clicks, but it should be fairly easy to implement and much faster than the current situation.

**Reply by larspalo** [reply]
This simple solution works surprisingly well in my tests. The question for this solution is if the removal message box that ask for confirmation the first time should be re-enabled by some condition or not? One could for instance re-enable it if all elements of that type was removed, or if another type is removed. Suggestions or ideas?

**Reply by larspalo** [reply]
I've committed a version of this simple way of speeding up the element deletion - or "Remove element" button handling. Somewhere down the road an other way of deleting multiple elements at once might be implemented, but I think I'll work with this for a while to get a feel for its usability. I already feel that it's at least an improvement!

**Comment by davidgritter** [comment]
I am happy with your approach to element deletion.  However, the same sort of  thing is needed for creating ranks, stops, and switches.  Recently I've been editing an ODF in which I need to add several hundred ranks and stops, and with the current method one goes to the Rank heading to add a rank, which puts the new rank at the bottom of the list of ranks, so that one has to scroll up through several pages of the rank list to add the next rank.  Would it be possible to  add multiple ranks at one time, or possibly add the new rank at the top of the list instead of the bottom(without re-arranging rank numbers)?  Or as you suggest right click on a rank in the element list and copy it.

**Reply by larspalo** [reply]
I'd guess that it should not be too much work to include a spinctrl on the base object type panel in front of the create button for creating one (default value) or many at once. If that would be good enough, I can start researching it...

**Reply by davidgritter** [reply]
That should work OK, the new items can all have the same name, since this appears to be allowed.  Then one can go to the bottom of the list and edit them individually without needing a lot of scrolling.

**Reply by davidgritter** [reply]
alternatively, you could set it up so that when you add a rank, switch, or stop, GOODF doesn't jump to the bottom of the list, but stays on the RANK, STOP, or SWITCH window so that its button can be pressed multiple times.

**Reply by larspalo** [reply]
I've implemented possibility to create (add) multiple windchestgroups, enclosures, ranks, switches and stops to begin with as the [Organ] presently can have up to 999 objects of each type. Other objects could possibly qualify too, but for now I think this is enough to get a feel for how it's working.

**Reply by davidgritter** [reply]
This makes life much easier, thank you

**Comment by Bumblebee001** [comment]
I imagine this situation would call for a Select button for every item and then (if selected) the options would be to copy, transfer, delete, paste, duplicate or whatever other task may be desireable.... Just throwing an idea in for brainstorming purposes.
